src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/openapi/TokensApiController.java

```diff
@@ -126,13 +139,26 @@ private Token getTokenFromService(String id) {
     @PreAuthorize("hasAuthority('EDIT_KEYTABLE_FRIENDLY_NAMES')")
     @Override
     public ResponseEntity<Token> updateToken(String id, TokenName tokenName) {
-        TokenInfo tokenInfo = null;
         try {
-            tokenInfo = tokenService.updateTokenFriendlyName(id, tokenName.getName());
-        } catch (TokenService.TokenNotFoundException e) {
+            TokenInfo tokenInfo = tokenService.updateTokenFriendlyName(id, tokenName.getName());
+            Token token = tokenConverter.convert(tokenInfo);
+            return new ResponseEntity<>(token, HttpStatus.OK);
+        } catch (TokenNotFoundException e) {
             throw new ResourceNotFoundException(e);
         }
-        Token token = tokenConverter.convert(tokenInfo);
-        return new ResponseEntity<>(token, HttpStatus.OK);
+    }
+
+    @PreAuthorize("hasAuthority('GENERATE_KEY')")
+    @Override
+    public ResponseEntity<Key> addKey(String tokenId, KeyLabel keyLabel) {
+        try {
+            KeyInfo keyInfo = keyService.addKey(tokenId, keyLabel.getLabel());
+            Key key = keyConverter.convert(keyInfo);
+            return new ResponseEntity<>(key, HttpStatus.OK);
```

Comment:
Instead of HTTP 200, should return 201 with some special headers, using ApiUtil.createCreatedResponse

Reply:
Oh yeah I had already forgot that! WIP

Reply:
Alrighty then: https://github.com/nordic-institute/X-Road/pull/263/files#diff-3cc2815ca4e4e2fe42be1e7394e4975fR157